core/classifiers/CodeClassifier/winnowing.py

```diff
@@ -0,0 +1,78 @@
+# coding: utf-8
+#
+# Copyright 2017 The Oppia Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Utility functions for using winnowing as pre-processing strp."""
+
+
+def hash_generator(token_to_id, tokens):
+    """Generate hash for tokens in 'tokens' using 'token_to_id'.
+
+    Args:
+        token_to_id: dict. A dictionary which maps each token to a unique ID.
+        tokens: list(str). A list of tokens.
+
+    Returns:
+        int. Hash value generated for tokens in 'tokens' using 'token_to_id'.
+    """
+    hash_val = 0
+    n = len(tokens) - 1
+    for x in tokens:
+        hash_val += token_to_id[x] * (len(token_to_id) ** n)
```

Comment:
No even theoretically it is not possible to have collision in this hash function. It is same as base conversion. For e.g. if you have 16 tokens (0 to 15) then what you would be doing here is converting hexadecimal value in its corresponding decimal value.

Let's say you have 16 tokens and they are same as hexadecimal digits (0 to 9, A to F). Now convert any substring of length 3 (consisting of these 16 tokens into a has value using above function). For e.g. consider A9D. It is a valid substring. Now what above function does is 
```
hash = id[A] * (16^2) + id[9] * (16^1) + id[D] * (16^0)
hash = 10 * (256) + 9 * 16 + 13 * 1)
hash = 2717
```
Check this for any substring of length 3, made up of (0 to 9, A to E) tokens. It is guaranteed to have unique, non repeated decimal value associated with it. You know it is true because every hexadecimal has unique decimal.

Above function is simply base conversion when input base is `len(tokens)` and output is decimal number.